lib/transformer.js

```diff
@@ -0,0 +1,158 @@
+'use strict'
+
+const zlib = require('zlib')
+const aws = require('aws-sdk')
+const {
+  checkFileFormat
+, formatObjectKey
+, hasProperty
+} = require('./utils.js')
+
+const s3 = new aws.S3({
+  apiVersion: '2006-03-01'
+})
+
+module.exports = {
+  extractData
+, getLogs
+, prepareLogs
+, s3
+}
+
+function extractData(data) {
+  if (!data || !Object.keys(data).length) return []
+
+  const {
+    Records
+  , logFiles
+  , ...meta
+  } = data
+
+  const base = {
+    timestamp: Date.now()
+  }
+
+  let records = data
+  if (Records) {
+    records = Records
+    base.meta = meta
+  } else if (logFiles) {
+    records = logFiles
+    base.meta = meta
+  }
+
+  if (meta && meta.eventTime) {
+    base.timestamp = (new Date(meta.eventTime)).getTime()
+  }
+
+  if (!Array.isArray(records)) {
+    const lineJSON = {
+      line: records
+    , ...base
+    }
+
+    if (typeof records !== 'string') {
+      lineJSON.line = JSON.stringify(records)
+    }
+
+    return [lineJSON]
+  }
+
+  return records.map((record) => {
+    const lineJSON = {
+      line: record
+    , ...base
+    }
+
+    if (typeof record !== 'string') {
+      lineJSON.line = JSON.stringify(record)
+    }
+
+    if (record.eventTime) {
+      lineJSON.timestamp = (new Date(record.eventTime)).getTime()
+    }
+
+    return lineJSON
+  })
+}
+
+async function getLogs(params) {
+  let err
+  if (!hasProperty(params, 'Key') || !hasProperty(params, 'Bucket')) {
+    err = new Error('Both Bucket and Key params must be provided')
+    err.meta = {params}
+    throw err
+  }
+
+  let data
+  try {
+    data = await s3.getObject(params).promise()
```

Comment:
@mdeltito @esatterwhite @darinspivey 

can anyone help me in mocking this? Mocking `s3.getObject` is simple but while testing on `AWS`, I have seen it requires `.promise()` too but having difficulty in testing

Thanks a lot in advance!

Reply:
Sure, I'll take a look.

Reply:
@smusali I fixed the issue in a new commit, so sync up with origin.  After reading [the SDK docs](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getObject-property), I couldn't find any reference to `.promise()`, in fact, all of these except 1 are callback-style.  Using build-style methods (chaining) for API calls makes it *really* hard/yucky to test, so we should stay away from those anyway.

The fix that's up in that commit changes the following:
* Use `promisify(s3.getObject)` and export that
* Call the promisified function from `exports` to facilitate testing
* Fixes the test to facilitate mocking - DO NOT destructure exported properties that you wish to mock, or else you can't mock them properly.
* Change the mock function to `async` to match the code
* Change the assertion to a `t.rejects` which is the proper way to test promise rejections (if you use try/catch, your assertion may not run if it doesn't throw due to a bug!)

I had some other comments regarding the `Error` itself, but I'll comment separately for that.

Reply:
Thanks, @darinspivey! Let's get this fixed first and I will modify the rest!